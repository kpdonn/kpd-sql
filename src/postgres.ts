import { SqlPart, LookupParamNum } from "./everything"

export function printSql(sqlPart: SqlPart, pNum: LookupParamNum): string {
  const pr = (it: SqlPart): string => {
    switch (it.sqlKind) {
      case "plainJoin":
        return `join ${pr(it.joinTable)} on ${pr(it.onCondition)}`
      case "leftJoin":
        return `left join ${pr(it.joinTable)} on ${pr(it.onCondition)}`
      case "table":
        return `${it._table} ${it._tableAs}`
      case "hardcoded":
      case "placeholderParam":
        return `$${pNum(it) + 1}`
      case "and":
        return `(${pr(it.left)}) AND (${pr(it.right)})`
      case "or":
        return `(${pr(it.left)}) OR (${pr(it.right)})`
      case "column":
        return `${it._tableAs}.${it._column}`
      case "columnDeclaration":
        return `${pr(it.col)} as ${it.col._columnAs}`
      case "eq":
        return `${pr(it.left)} = ${pr(it.right)}`
      case "selectStatement":
        const columnsSql = it.columns.map(pr).join(", ")
        const fromSql = it.fromTables.map(pr).join(", ")
        const joinSql = it.joins.map(pr).join("\n")
        const whereSql = it.where ? `WHERE ${pr(it.where)}` : ""
        return `SELECT ${columnsSql} FROM ${fromSql} ${joinSql} ${whereSql}`
    }
  }
  return pr(sqlPart)
}
