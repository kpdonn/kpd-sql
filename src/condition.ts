import { Literal, tblAsSym, ColType, tySym, SqlParam, SqlParamName } from "./utils"
import { Column } from "./table"

export const condTbls: unique symbol = Symbol("conditionTables")
export type cts = typeof condTbls
export const cp: unique symbol = Symbol("conditionParams")
export type cps = typeof cp

export type ConditionType = EqCondition | AndCondition | OrCondition

export abstract class Condition<CT extends string = string, P = {}> {
  readonly [condTbls]: CT;
  readonly [cp]: P
  abstract readonly kind: Literal<ConditionType["kind"]>

  and<C extends Condition>(cond: C): Condition<CT | C[cts], P & C[cps]> {
    return new AndCondition(this, cond)
  }
  or<C extends Condition>(cond: C): Condition<CT | C[cts], P & C[cps]> {
    return new OrCondition(this, cond)
  }
}

export class AndCondition<CT extends string = string, P = {}> extends Condition<CT, P> {
  readonly kind = "and"
  constructor(readonly left: Condition, readonly right: Condition) {
    super()
  }
}

export class OrCondition<CT extends string = string, P = {}> extends Condition<CT, P> {
  readonly kind = "or"

  constructor(readonly left: Condition, readonly right: Condition) {
    super()
  }
}

export class EqCondition<
  Col1 extends Column = Column,
  Col2 extends Column<string, Col1[tySym]> = never,
  SPN extends string = never
> extends Condition<
  Col1[tblAsSym] | Literal<Col2[tblAsSym]>,
  SqlParam<SPN, ColType<Col1>>
> {
  readonly kind = "eq"

  constructor(
    readonly left: Col1,
    readonly right: Col2 | ColType<Col1> | SqlParamName<SPN>
  ) {
    super()
  }
}
