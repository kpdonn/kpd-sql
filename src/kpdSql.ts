import * as t from "io-ts"
import {
  tblSym,
  colSym,
  tySym,
  col,
  ty,
  tblAsSym,
  tblAs,
  tbl,
  colAsSym,
  colAs,
} from "./implementation"
import { Table, ColInfo, Condition, Literal } from "./table"

export type ArrayKeys = keyof any[]

export type TupleKeys<T> = Exclude<keyof T, ArrayKeys>

export type TabCols<
  ColArr extends { [index: string]: ColInfo },
  ArrInd extends keyof ColArr,
  TblNames extends string
> = { [Ind in ArrInd]: ColArr[Ind][tblAsSym] extends TblNames ? Ind : never }[ArrInd]

export type Grab<
  ColArr extends { [index: string]: ColInfo },
  ArrInd extends keyof ColArr,
  TblNames extends string
> = [TabCols<ColArr, ArrInd, TblNames>] extends [never]
  ? {}
  : GetVal<ColArr[TabCols<ColArr, ArrInd, TblNames>][colAsSym], ColArr>

export type GetVal<
  ColNames extends string,
  ColArr extends { [index: string]: ColInfo }
> = { [K in ColNames]: FilterVal<K, ColArr[keyof ColArr]> }

export type TypeCol<N extends string, U extends t.Any> = { [colAs]: N; [ty]: U }

export type FilterVal<N extends string, T extends ColInfo> = T extends TypeCol<N, infer U>
  ? t.TypeOf<U>
  : never

export interface SQLFrom<
  Tables extends string = never,
  OptTables extends string = never,
  Cols = {},
  P = {}
> {
  from<T extends Table & NoDupTable<T, Tables>>(
    table: T
  ): SQLJoin<Tables | T[tblAsSym], OptTables, Cols, P>
}

export type GCol<C extends ColInfo> = C
export type SafeInd<T, K extends keyof Base, Base = T> = Extract<T, Base> extends never
  ? never
  : Extract<T, Base>[K]

export type NoDuplicates<T extends string> = { "NO DUPLICATE KEYS ALLOWED": T }
export type ColumnsObj<T> = { [K in TupleKeys<T>]: Extract<T[K], ColInfo> }

export interface SQLColumns<
  RT extends string,
  OT extends string,
  Cols = {},
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> {
  columns<
    C extends {
      [K in keyof T]: T[K][colAsSym] extends (
        | (SafeInd<T[Exclude<keyof T, K>], colAsSym>)
        | keyof Cols)
        ? NoDuplicates<SafeInd<T[K], colAsSym>>
        : ColInfo<TblNames>
    } & { "0": any },
    T extends ColumnsObj<C> = ColumnsObj<C>
  >(
    cols: C
  ): SQLSelectAndWhere<
    RT,
    OT,
    Cols &
      Grab<C, TupleKeys<C>, Literal<RT>> &
      EmptyIfNone<Partial<Grab<C, TupleKeys<C>, Literal<OT>>>>,
    P
  >
}

export type EmptyIfNone<T> = string extends keyof T
  ? {}
  : [keyof T] extends [never] ? {} : T

export interface SQLSelectAndWhere<RT extends string, OT extends string, Cols, P = {}>
  extends SQLColumns<RT, OT, Cols, P>,
    SQLWhere<RT, OT, Cols, P> {}

export type NoDupTable<T extends Table, TableNames extends string> = {
  [tbl]: T[tblAsSym] extends TableNames ? never : string
}
export interface SQLJoin<
  RT extends string,
  OT extends string,
  Cols,
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> extends SQLColumns<RT, OT, Cols, P> {
  join<T extends Table & NoDupTable<T, TblNames>, LCol extends ColInfo<TblNames>>(
    table: T,
    cond: Condition<TblNames | T[tblAsSym]>
  ): SQLJoin<RT | T[tblAsSym], OT, Cols, P>

  leftJoin<T extends Table & NoDupTable<T, TblNames>, LCol extends ColInfo<TblNames>>(
    table: T,
    cond: Condition<TblNames | T[tblAsSym]>
  ): SQLJoin<RT, OT | T[tblAsSym], Cols, P>
}

export type ExecuteArgs<P, Cols> = {} extends P
  ? () => Promise<Cols[]>
  : (args: P) => Promise<Cols[]>

export interface SQLExecute<RT extends string, OT extends string, Cols, P = {}>
  extends SQLReady<Cols> {
  execute: ExecuteArgs<P, Cols>
  toSql(): string
}

export declare const sqlReady: unique symbol
export interface SQLReady<Cols> {
  [sqlReady]: Cols
}

export interface SQLWhere<
  RT extends string,
  OT extends string,
  Cols,
  P = {},
  TblNames extends string = Literal<RT> | Literal<OT>
> extends SQLExecute<RT, OT, Cols, P> {
  where<CTbles extends TblNames, SP>(
    cond: Condition<CTbles, SP>
  ): SQLWhere<RT, OT, Cols, P & SP>

  whereSub<CTbles extends TblNames, SP>(
    cond: (subSelect: SQLFrom<RT, OT, {}>) => Condition<CTbles, SP>
  ): SQLWhere<RT, OT, Cols, P & SP>
}
