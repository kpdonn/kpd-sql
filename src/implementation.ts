import * as Immutable from "immutable"
import { Pool } from "pg"
import { TableWithColumns } from "./utils"
import { Column } from "./table"
import { Condition } from "./condition"
import { SqlBuilder } from "./kpdSql"

let dbPool: Pool

export function initializePool(pool: Pool): void {
  dbPool = pool
}

interface Join {
  type: "inner" | "left"
  table: TableWithColumns
  cond: Condition<any>
}

const BuilderState = Immutable.Record({
  fromTable: (undefined as any) as TableWithColumns,
  joins: Immutable.List<Join>(),
  columns: Immutable.List(),
  wheres: Immutable.List(),
})
class SqlBuilderImpl {
  constructor(private readonly state = BuilderState()) {}

  from(table: TableWithColumns): SqlBuilderImpl {
    let newState = this.state.withMutations(arg => {
      arg.set("fromTable", table)
    })

    return new SqlBuilderImpl(newState)
  }

  join(table: TableWithColumns, cond: Condition<any>): SqlBuilderImpl {
    return this.addJoin("inner", table, cond)
  }

  leftJoin(table: TableWithColumns, cond: Condition<any>): SqlBuilderImpl {
    return this.addJoin("left", table, cond)
  }

  addJoin(
    type: "inner" | "left",
    table: TableWithColumns,
    cond: Condition<any>
  ): SqlBuilderImpl {
    const newState = this.state.updateIn(["joins"], joins =>
      joins.push({ type, table, cond })
    )

    return new SqlBuilderImpl(newState)
  }

  columns(cols: Column[]): SqlBuilderImpl {
    const newState = this.state.updateIn(["columns"], stateCols =>
      stateCols.withMutations((mut: any) => {
        cols.forEach(col => mut.push(col))
      })
    )
    return new SqlBuilderImpl(newState)
  }

  where(cond: Condition<any>): SqlBuilderImpl {
    const newState = this.state.updateIn(["wheres"], wheres => wheres.push({ cond }))

    return new SqlBuilderImpl(newState)
  }

  toSql(): string {
    let sql = "Select "
    const state = this.state
    const selectCols = state.columns.map(col => col.toSql()).join(", ")

    sql += selectCols + " \n"

    const fromClause = ` from ${state.fromTable} `

    sql += fromClause

    state.joins.forEach(join => {
      sql += " \n "
      const onClause = ` on ${join.cond} `
      sql += ` ${join.type} join ${join.table} ${onClause} `
    })

    sql += "\n where "

    state.wheres.forEach(where => {
      sql += ` ${where.cond.toSql()} `
    })

    return sql
  }

  execute(): Promise<any> {
    const sql = this.toSql()

    const result = dbPool.query(sql).then(res => res.rows)

    return result
  }
}

export function select(): SqlBuilder {
  return new SqlBuilderImpl() as any
}

export function tuple<T extends any[] & { "0": any }>(array: T): T {
  return array
}
