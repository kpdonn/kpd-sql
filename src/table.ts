import * as t from "io-ts"
import { ty, tbl, col, tySym, tblSym } from "./implementation"

export declare const condTbls: unique symbol
export type condTblsSym = typeof condTbls

export type InCol<CN extends string = string, Type extends t.Any = t.Mixed> = Record<
  CN,
  { type: Type }
>

export type ColInfo<
  TN extends string = string,
  Type extends t.Any = t.Mixed,
  CN extends string = string
> = {
  [ty]: Type
  [tbl]: TN
  [col]: CN
}

export type OutCol<
  TN extends string = string,
  Type extends t.Any = t.Mixed,
  CN extends string = string,
  ThisCol extends ColInfo = ColInfo<TN, Type, CN>
> = ThisCol & { as: AsCol<Type, TN> } & Comparisons<ThisCol>

export type AsCol<Type extends t.Any, TN extends string> = <NN extends string>(
  newName: NN
) => OutCol<TN, Type, NN>

export type TransformInCol<TN extends string, C extends InCol> = {
  [K in keyof C]: OutCol<TN, C[K]["type"], K>
}

export type Table<TN extends string = string> = {
  [tbl]: TN
}

export type TableOut<TN extends string, C extends InCol> = Table<TN> &
  TransformInCol<TN, C>

export function table<N extends string, C extends InCol>(arg: {
  name: N
  columns: C
}): TableOut<N, C> {
  const result: any = { [tbl]: arg.name }

  Object.keys(arg.columns).forEach(colName => {
    result[colName] = {
      [tbl]: arg.name,
      name: colName,
      ...arg.columns[colName]
    }
  })

  return result
}
export type Literal<T extends string> = string extends T ? never : T
export interface Condition<TblNames extends string> {
  [condTbls]: TblNames

  and<C extends Condition<any>>(cond: C): Condition<TblNames | C[condTblsSym]>
  or<C extends Condition<any>>(cond: C): Condition<TblNames | C[condTblsSym]>
}

export type Comparisons<Col1 extends ColInfo> = {
  eq: <Col2 extends ColInfo<any, Col1[tySym]>>(
    col2: Col2 | t.TypeOf<Col1[tySym]>
  ) => Condition<Col1[tblSym] | Literal<Col2[tblSym]>>
}

const Book = table({
  name: "book",
  columns: {
    id: { type: t.number },
    title: { type: t.string },
    year: { type: t.number },
    pages: { type: t.number },
    authorId: { type: t.number }
  }
})

const comp = Book.id.eq(Book.year)
