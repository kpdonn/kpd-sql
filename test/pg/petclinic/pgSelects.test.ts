import * as fs from "fs"
import * as path from "path"
import { Pool } from "pg"
import { init, param } from "../../../src/everything"
import { PgPlugin } from "../../../src/postgres"
import { Owner, Pet, Specialty, Type, Vet, VetSpecialty, Visit } from "./tables"

const connectionString = "postgresql://localhost:5432/sqltest"

const pool = new Pool({
  connectionString,
})

const db = init(PgPlugin.init(pool))

describe("Select query tests", () => {
  it("select pets seen dr Ortega", async () => {
    const query = db
      .select()
      .from(Pet.join(Visit, Pet.id.eq(Visit.petId)).join(Vet, Vet.id.eq(Visit.vetId)))
      .columns([Pet.id, Pet.name, Pet.birthDate, Pet.ownerId, Pet.typeId])
      .where(Vet.lastName.eq(param("vetLastName")))

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute({ vetLastName: "Ortega" })
    expect(results).toHaveLength(2)
    expect(results).toMatchSnapshot()
  })

  it("select with subquery pets live in madison", async () => {
    const query = db
      .select()
      .from(Pet)
      .columns([Pet.id, Pet.name, Pet.birthDate, Pet.ownerId, Pet.typeId])
      .where(sq => {
        return Pet.ownerId.in(
          sq
            .from(Owner)
            .columns([Owner.id])
            .where(Owner.city.eq(param("ownerCity")))
        )
      })

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute({ ownerCity: "Madison" })
    expect(results).toHaveLength(4)
    expect(results).toMatchSnapshot()
  })

  it("select pets hardcoded live in madison or McFarland", async () => {
    const query = db
      .select()
      .from(Pet)
      .columns([Pet.id, Pet.name, Pet.birthDate, Pet.ownerId, Pet.typeId])
      .where(sq => {
        return Pet.ownerId.in(
          sq
            .from(Owner)
            .columns([Owner.id])
            .where(Owner.city.in(["Madison", "McFarland"]))
        )
      })

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute()
    expect(results).toHaveLength(6)
    expect(results).toMatchSnapshot()
  })

  it("select non cat pets with optional visit", async () => {
    const query = db
      .select()
      .from(
        Pet.join(Type, Type.id.eq(Pet.typeId)).leftJoin(Visit, Pet.id.eq(Visit.petId))
      )
      .columns([
        Pet.id,
        Type.name.as("type"),
        Pet.name,
        Pet.birthDate,
        Pet.ownerId,
        Pet.typeId,
        Visit.visitDate,
        Visit.vetId,
      ])
      .where(Type.name.not.eq("cat"))

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute()
    expect(results).toHaveLength(9)
    expect(results).toMatchSnapshot()
  })

  it("select vets with no specialty", async () => {
    const query = db
      .select()
      .from(
        Vet.leftJoin(VetSpecialty, VetSpecialty.vetId.eq(Vet.id)).leftJoin(
          Specialty,
          Specialty.id.eq(VetSpecialty.specialtyId)
        )
      )

      .columns([Vet.id, Vet.firstName, Vet.lastName, Specialty.name])
      .where(VetSpecialty.specialtyId.isNull)

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute()
    expect(results).toHaveLength(2)
    expect(results).toMatchSnapshot()
  })

  it("select vets with their specialties", async () => {
    const query = db
      .select()
      .from(
        Vet.leftJoin(VetSpecialty, VetSpecialty.vetId.eq(Vet.id)).leftJoin(
          Specialty,
          Specialty.id.eq(VetSpecialty.specialtyId)
        )
      )

      .columns([Vet.id, Vet.firstName, Vet.lastName, Specialty.name])
      .where(VetSpecialty.specialtyId.isNotNull)

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute()
    expect(results).toHaveLength(5)
    expect(results).toMatchSnapshot()
  })

  it("with clause query", async () => {
    const query = db
      .with(
        "myWith",
        db
          .from(
            Vet.leftJoin(VetSpecialty, VetSpecialty.vetId.eq(Vet.id)).leftJoin(
              Specialty,
              Specialty.id.eq(VetSpecialty.specialtyId)
            )
          )
          .columns([Vet.id, Vet.firstName, Vet.lastName, Specialty.name])
          .where(VetSpecialty.specialtyId.isNotNull)
      )
      .select(sq =>
        sq
          .from(sq.table.myWith.join(Visit, Visit.vetId.eq(sq.table.myWith.id)))
          .columns([sq.table.myWith.id, sq.table.myWith.lastName, Visit.petId])
      )

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute()
    expect(results).toHaveLength(3)
    expect(results).toMatchSnapshot()
  })

  it("with clause cats had visit with surgeon", async () => {
    const query = db
      .with(
        "myWith",
        db
          .from(
            Vet.leftJoin(VetSpecialty, VetSpecialty.vetId.eq(Vet.id)).leftJoin(
              Specialty,
              Specialty.id.eq(VetSpecialty.specialtyId)
            )
          )
          .columns([Vet.id, Vet.firstName, Vet.lastName, Specialty.name])
          .where(Specialty.name.eq(param("specialty")))
      )
      .select(sq =>
        sq
          .from(
            sq.table.myWith.leftJoin(
              Visit.join(Pet.join(Type, Pet.typeId.eq(Type.id)), Visit.petId.eq(Pet.id)),
              Visit.vetId.eq(sq.table.myWith.id)
            )
          )
          .columns([sq.table.myWith.id, sq.table.myWith.lastName, Visit.petId, Pet.name])
          .where(sq.table.myWith.lastName.not.eq("FakeLastName").and(Type.name.eq("cat")))
      )

    expect(query.toSql()).toMatchSnapshot()
    const results = await query.execute({ specialty: "surgery" })
    expect(results).toHaveLength(2)
    expect(results).toMatchSnapshot()
  })
})

beforeAll(async () => {
  const initDb = fs.readFileSync(
    path.resolve(__dirname, "postgres-petclinic-init.sql"),
    "utf8"
  )

  await pool.query(initDb)
})

afterAll(async () => {
  await pool.end()
})
