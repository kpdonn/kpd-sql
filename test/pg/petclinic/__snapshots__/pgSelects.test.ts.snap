// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select query tests select non cat pets with optional visit 1`] = `
"SELECT
\\"pet\\".\\"id\\" as \\"id\\",
\\"type\\".\\"name\\" as \\"type\\",
\\"pet\\".\\"name\\" as \\"name\\",
\\"pet\\".\\"birth_date\\" as \\"birthDate\\",
\\"pet\\".\\"owner_id\\" as \\"ownerId\\",
\\"pet\\".\\"type_id\\" as \\"typeId\\",
\\"visit\\".\\"visit_date\\" as \\"visitDate\\",
\\"visit\\".\\"vet_id\\" as \\"vetId\\"
FROM
( ( \\"pet\\" JOIN \\"type\\" ON \\"type\\".\\"id\\" = \\"pet\\".\\"type_id\\" ) LEFT JOIN \\"visit\\" ON \\"pet\\".\\"id\\" = \\"visit\\".\\"pet_id\\" )
WHERE NOT (\\"type\\".\\"name\\" = $1)"
`;

exports[`Select query tests select non cat pets with optional visit 2`] = `
Array [
  anonymous {
    "birthDate": 1997-02-24T06:00:00.000Z,
    "id": 10,
    "name": "Mulligan",
    "ownerId": 8,
    "type": "dog",
    "typeId": 2,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2000-01-20T06:00:00.000Z,
    "id": 6,
    "name": "George",
    "ownerId": 5,
    "type": "snake",
    "typeId": 4,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2000-03-09T06:00:00.000Z,
    "id": 11,
    "name": "Freddy",
    "ownerId": 9,
    "type": "bird",
    "typeId": 5,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2000-06-24T05:00:00.000Z,
    "id": 12,
    "name": "Lucky",
    "ownerId": 10,
    "type": "dog",
    "typeId": 2,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2002-08-06T05:00:00.000Z,
    "id": 2,
    "name": "Basil",
    "ownerId": 2,
    "type": "hamster",
    "typeId": 6,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2000-11-30T06:00:00.000Z,
    "id": 5,
    "name": "Iggy",
    "ownerId": 4,
    "type": "lizard",
    "typeId": 3,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2000-03-07T06:00:00.000Z,
    "id": 4,
    "name": "Jewel",
    "ownerId": 3,
    "type": "dog",
    "typeId": 2,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 2001-04-17T05:00:00.000Z,
    "id": 3,
    "name": "Rosy",
    "ownerId": 3,
    "type": "dog",
    "typeId": 2,
    "vetId": null,
    "visitDate": null,
  },
  anonymous {
    "birthDate": 1999-08-06T05:00:00.000Z,
    "id": 9,
    "name": "Lucky",
    "ownerId": 7,
    "type": "bird",
    "typeId": 5,
    "vetId": null,
    "visitDate": null,
  },
]
`;

exports[`Select query tests select pets hardcoded live in madison or McFarland 1`] = `
"SELECT
\\"pet\\".\\"id\\" as \\"id\\",
\\"pet\\".\\"name\\" as \\"name\\",
\\"pet\\".\\"birth_date\\" as \\"birthDate\\",
\\"pet\\".\\"owner_id\\" as \\"ownerId\\",
\\"pet\\".\\"type_id\\" as \\"typeId\\"
FROM
\\"pet\\"
WHERE \\"pet\\".\\"owner_id\\" = ANY (SELECT
\\"owner\\".\\"id\\" as \\"id\\"
FROM
\\"owner\\"
WHERE \\"owner\\".\\"city\\" = ANY ($1))"
`;

exports[`Select query tests select pets hardcoded live in madison or McFarland 2`] = `
Array [
  anonymous {
    "birthDate": 2000-09-07T05:00:00.000Z,
    "id": 1,
    "name": "Leo",
    "ownerId": 1,
    "typeId": 1,
  },
  anonymous {
    "birthDate": 2001-04-17T05:00:00.000Z,
    "id": 3,
    "name": "Rosy",
    "ownerId": 3,
    "typeId": 2,
  },
  anonymous {
    "birthDate": 2000-03-07T06:00:00.000Z,
    "id": 4,
    "name": "Jewel",
    "ownerId": 3,
    "typeId": 2,
  },
  anonymous {
    "birthDate": 2000-01-20T06:00:00.000Z,
    "id": 6,
    "name": "George",
    "ownerId": 5,
    "typeId": 4,
  },
  anonymous {
    "birthDate": 1997-02-24T06:00:00.000Z,
    "id": 10,
    "name": "Mulligan",
    "ownerId": 8,
    "typeId": 2,
  },
  anonymous {
    "birthDate": 2000-03-09T06:00:00.000Z,
    "id": 11,
    "name": "Freddy",
    "ownerId": 9,
    "typeId": 5,
  },
]
`;

exports[`Select query tests select pets seen dr Ortega 1`] = `
"SELECT
\\"pet\\".\\"id\\" as \\"id\\",
\\"pet\\".\\"name\\" as \\"name\\",
\\"pet\\".\\"birth_date\\" as \\"birthDate\\",
\\"pet\\".\\"owner_id\\" as \\"ownerId\\",
\\"pet\\".\\"type_id\\" as \\"typeId\\"
FROM
( ( \\"pet\\" JOIN \\"visit\\" ON \\"pet\\".\\"id\\" = \\"visit\\".\\"pet_id\\" ) JOIN \\"vet\\" ON \\"vet\\".\\"id\\" = \\"visit\\".\\"vet_id\\" )
WHERE \\"vet\\".\\"last_name\\" = $1"
`;

exports[`Select query tests select pets seen dr Ortega 2`] = `
Array [
  anonymous {
    "birthDate": 1995-09-04T05:00:00.000Z,
    "id": 8,
    "name": "Max",
    "ownerId": 6,
    "typeId": 1,
  },
  anonymous {
    "birthDate": 1995-09-04T05:00:00.000Z,
    "id": 7,
    "name": "Samantha",
    "ownerId": 6,
    "typeId": 1,
  },
]
`;

exports[`Select query tests select vets with no specialty 1`] = `
"SELECT
\\"vet\\".\\"id\\" as \\"id\\",
\\"vet\\".\\"first_name\\" as \\"firstName\\",
\\"vet\\".\\"last_name\\" as \\"lastName\\",
\\"specialty\\".\\"name\\" as \\"name\\"
FROM
( ( \\"vet\\" LEFT JOIN \\"vet_specialty\\" ON \\"vet_specialty\\".\\"vet_id\\" = \\"vet\\".\\"id\\" ) LEFT JOIN \\"specialty\\" ON \\"specialty\\".\\"id\\" = \\"vet_specialty\\".\\"specialty_id\\" )
WHERE \\"vet_specialty\\".\\"specialty_id\\" IS NULL"
`;

exports[`Select query tests select vets with no specialty 2`] = `
Array [
  anonymous {
    "firstName": "Sharon",
    "id": 6,
    "lastName": "Jenkins",
    "name": null,
  },
  anonymous {
    "firstName": "James",
    "id": 1,
    "lastName": "Carter",
    "name": null,
  },
]
`;

exports[`Select query tests select vets with their specialties 1`] = `
"SELECT
\\"vet\\".\\"id\\" as \\"id\\",
\\"vet\\".\\"first_name\\" as \\"firstName\\",
\\"vet\\".\\"last_name\\" as \\"lastName\\",
\\"specialty\\".\\"name\\" as \\"name\\"
FROM
( ( \\"vet\\" LEFT JOIN \\"vet_specialty\\" ON \\"vet_specialty\\".\\"vet_id\\" = \\"vet\\".\\"id\\" ) LEFT JOIN \\"specialty\\" ON \\"specialty\\".\\"id\\" = \\"vet_specialty\\".\\"specialty_id\\" )
WHERE \\"vet_specialty\\".\\"specialty_id\\" IS NOT NULL"
`;

exports[`Select query tests select vets with their specialties 2`] = `
Array [
  anonymous {
    "firstName": "Helen",
    "id": 2,
    "lastName": "Leary",
    "name": "radiology",
  },
  anonymous {
    "firstName": "Linda",
    "id": 3,
    "lastName": "Douglas",
    "name": "surgery",
  },
  anonymous {
    "firstName": "Linda",
    "id": 3,
    "lastName": "Douglas",
    "name": "dentistry",
  },
  anonymous {
    "firstName": "Rafael",
    "id": 4,
    "lastName": "Ortega",
    "name": "surgery",
  },
  anonymous {
    "firstName": "Henry",
    "id": 5,
    "lastName": "Stevens",
    "name": "radiology",
  },
]
`;

exports[`Select query tests select with subquery pets live in madison 1`] = `
"SELECT
\\"pet\\".\\"id\\" as \\"id\\",
\\"pet\\".\\"name\\" as \\"name\\",
\\"pet\\".\\"birth_date\\" as \\"birthDate\\",
\\"pet\\".\\"owner_id\\" as \\"ownerId\\",
\\"pet\\".\\"type_id\\" as \\"typeId\\"
FROM
\\"pet\\"
WHERE \\"pet\\".\\"owner_id\\" = ANY (SELECT
\\"owner\\".\\"id\\" as \\"id\\"
FROM
\\"owner\\"
WHERE \\"owner\\".\\"city\\" = $1)"
`;

exports[`Select query tests select with subquery pets live in madison 2`] = `
Array [
  anonymous {
    "birthDate": 2000-09-07T05:00:00.000Z,
    "id": 1,
    "name": "Leo",
    "ownerId": 1,
    "typeId": 1,
  },
  anonymous {
    "birthDate": 2000-01-20T06:00:00.000Z,
    "id": 6,
    "name": "George",
    "ownerId": 5,
    "typeId": 4,
  },
  anonymous {
    "birthDate": 1997-02-24T06:00:00.000Z,
    "id": 10,
    "name": "Mulligan",
    "ownerId": 8,
    "typeId": 2,
  },
  anonymous {
    "birthDate": 2000-03-09T06:00:00.000Z,
    "id": 11,
    "name": "Freddy",
    "ownerId": 9,
    "typeId": 5,
  },
]
`;

exports[`Select query tests with clause cats had visit with surgeon 1`] = `
"WITH \\"myWith\\" AS (SELECT
\\"vet\\".\\"id\\" as \\"id\\",
\\"vet\\".\\"first_name\\" as \\"firstName\\",
\\"vet\\".\\"last_name\\" as \\"lastName\\",
\\"specialty\\".\\"name\\" as \\"name\\"
FROM
( ( \\"vet\\" LEFT JOIN \\"vet_specialty\\" ON \\"vet_specialty\\".\\"vet_id\\" = \\"vet\\".\\"id\\" ) LEFT JOIN \\"specialty\\" ON \\"specialty\\".\\"id\\" = \\"vet_specialty\\".\\"specialty_id\\" )
WHERE \\"specialty\\".\\"name\\" = $1)
SELECT
\\"myWith\\".\\"id\\" as \\"id\\",
\\"myWith\\".\\"lastName\\" as \\"lastName\\",
\\"visit\\".\\"pet_id\\" as \\"petId\\",
\\"pet\\".\\"name\\" as \\"name\\"
FROM
( \\"myWith\\" LEFT JOIN ( \\"visit\\" JOIN ( \\"pet\\" JOIN \\"type\\" ON \\"pet\\".\\"type_id\\" = \\"type\\".\\"id\\" ) ON \\"visit\\".\\"pet_id\\" = \\"pet\\".\\"id\\" ) ON \\"visit\\".\\"vet_id\\" = \\"myWith\\".\\"id\\" )
WHERE (NOT (\\"myWith\\".\\"lastName\\" = $2)) AND (\\"type\\".\\"name\\" = $3)"
`;

exports[`Select query tests with clause cats had visit with surgeon 2`] = `
Array [
  anonymous {
    "id": 4,
    "lastName": "Ortega",
    "name": "Max",
    "petId": 8,
  },
  anonymous {
    "id": 4,
    "lastName": "Ortega",
    "name": "Samantha",
    "petId": 7,
  },
]
`;

exports[`Select query tests with clause query 1`] = `
"WITH \\"myWith\\" AS (SELECT
\\"vet\\".\\"id\\" as \\"id\\",
\\"vet\\".\\"first_name\\" as \\"firstName\\",
\\"vet\\".\\"last_name\\" as \\"lastName\\",
\\"specialty\\".\\"name\\" as \\"name\\"
FROM
( ( \\"vet\\" LEFT JOIN \\"vet_specialty\\" ON \\"vet_specialty\\".\\"vet_id\\" = \\"vet\\".\\"id\\" ) LEFT JOIN \\"specialty\\" ON \\"specialty\\".\\"id\\" = \\"vet_specialty\\".\\"specialty_id\\" )
WHERE \\"vet_specialty\\".\\"specialty_id\\" IS NOT NULL)
SELECT
\\"myWith\\".\\"id\\" as \\"id\\",
\\"myWith\\".\\"lastName\\" as \\"lastName\\",
\\"visit\\".\\"pet_id\\" as \\"petId\\"
FROM
( \\"myWith\\" JOIN \\"visit\\" ON \\"visit\\".\\"vet_id\\" = \\"myWith\\".\\"id\\" )"
`;

exports[`Select query tests with clause query 2`] = `
Array [
  anonymous {
    "id": 2,
    "lastName": "Leary",
    "petId": 8,
  },
  anonymous {
    "id": 4,
    "lastName": "Ortega",
    "petId": 8,
  },
  anonymous {
    "id": 4,
    "lastName": "Ortega",
    "petId": 7,
  },
]
`;
